rpc: Tidy up reporting of buried and ongoing softforks

maflcko opened the pull request:
This unifies how softforks (e.g. buried, ISM, BIP9, ...) are reported. The output of `getforkinfo()` is now always a dict, where the key is the name of the softfork and the value is a dict with further details.

jnewbery commented:
utACK fa3d0199a8a03cc7e14b0242d672af0328f201f8

I've read the code and it looks good to me.

laanwj commented:
How does this fit in with buried softfork changes like  #16060?

jnewbery commented:
> How does this fit in with buried softfork changes like

This is the first commit from #16060, with some changes to the tests and release notes.

MarcoFalke commented:
Marking WIP

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1318,29 +1311,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "        }\n"
+            "      },\n"
+            "      \"height\": \"xxxxxx\",       (numeric) height of the first block which the rules are enforced (only for \"buried\" type, or \"bip9\" type with \"locked_in\" or \"active\" status)\n"
+            "      \"active\": xx,             (boolean) true if the rules are enforced\n"
```
note: Missing test for this when the type is "bip9"

laanwj commented:
>  This is the first commit from #16060, with some changes to the tests and release notes.

Thanks. Yes, probably makes sense to split this off from the burying itself.

ajtowns committed:
drop VersionBitsTip, explain versionbitscache guard

ajtowns reviewed (changes requested):
I think the "active" field needs fixing one way or the other before merge.

Three other suggestions:

 * include a `"bips": [141, 143, 147]` field to make it easy to look up what's being implemented
 * change the old names to "heightincb", "strictder" and "cltv" for consistency (and so you don't have to remember what bip34 was or which was bip65 and which was bip66)
 * move this into a "getforkinfo" rpc and out of "getblockchaininfo"; that way getblockchaininfo is just ~14 lines, and the bits that change every block aren't hidden behind a wall of soft fork info that rarely changes at all

Draft implementation of the above suggestions at https://github.com/ajtowns/bitcoin/commits/201907-getforkinfo if any of them seem worthwhile.

Edited to add: big concept ack!

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1385,16 +1372,14 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
+    BuriedForkDescPushBack(softforks, "bip65", consensusParams.BIP65Height, *tip);
     for (int pos = Consensus::DEPLOYMENT_CSV; pos != Consensus::MAX_VERSION_BITS_DEPLOYMENTS; ++pos) {
-        BIP9SoftForkDescPushBack(bip9_softforks, consensusParams, static_cast<Consensus::DeploymentPos>(pos));
+        BIP9SoftForkDescPushBack(softforks, consensusParams, static_cast<Consensus::DeploymentPos>(pos));
```
I don't think reverting this to a for loop makes things any clearer. There are just three bip9 softforks currently, so listing them out explicitly seems clearer to me.

MarcoFalke replied:
Refactoring changes should (if possible) be in a separate commit from behaviour changes. This can be done in a separate pull request

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1224,54 +1224,34 @@ static UniValue verifychain(const JSONRPCRequest& request)
+    softforks.pushKV(name, rv);
 }
 
 static UniValue BIP9SoftForkDesc(const Consensus::Params& consensusParams, Consensus::DeploymentPos id)
```
Again, I don't understand why you've reverted this from the tidier code in 16060

MarcoFalke replied:
Refactoring changes should (if possible) be in a separate commit from behaviour changes. This can be done in a separate pull request

jnewbery reviewed (commented):
A couple of comments inline

Incidentally, you've force-pushed to this branch 9 times since opening it (7 since I last reviewed), without any additional comments in the PR. Without any explanation for why you're making those changes, it was a lot of effort to re-review this PR. I don't see the rationale for reverting some of the previous changes, for example.

ajtowns commented on src/rpc/blockchain.cpp:
```diff
@@ -1318,29 +1309,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "        }\n"
+            "      },\n"
+            "      \"height\": \"xxxxxx\",       (numeric) height of the first block which the rules are enforced (only for \"buried\" type, or \"bip9\" type with \"locked_in\" or \"active\" status)\n"
+            "      \"active\": xx,             (boolean) true if the rules are enforced at the current chain tip\n"
```
After sleeping on it, I think I've changed my mind -- it's probably less confusing to have "active=true" tell you whether the rules were enforced on the current block, and you can use the "height=N" output to tell you whether it needs to be enforced on the next block.

In that case, I think you'd want `nHeight >= BIPxxHeight` and to use `tip->pprev` instead of `tip` for the versionbits state. The versionbits statistics function might need some more tweaking to make sense though.

ajtowns replied:
I've pushed an update to https://github.com/ajtowns/bitcoin/commits/201907-getforkinfo that implements the latter in a way I think makes sense, including changes to versionist statistics. Can check the output makes sense with something like (assuming the "csv" softfork activates at block 1152 and you have jshon available):

    $ for a in 863 864 865 1006 1007 1008 1009 1010 1150 1151 1152; do echo $a; bitcoin-cli -regtest getforkinfo $(bitcoin-cli -regtest getblockhash $a)  | jshon -e csv; echo ======; done | less

practicalswift commented on doc/release-notes-16328.md:
```diff
@@ -0,0 +1,9 @@
+
+- The array `softforks` and object `bip9_softforks` in the `getblockchaininfo`
+  RPC are combined into a single `softforks` object. Refer to the RPC help for
+  futher details.
```
Should be "further" :-)

Commit:
[rpc] Tidy up reporting of buried and ongoing softforks

This combines reporting of buried (formally ISM) softfork deployments
and BIP9 versionbits softfork deployments into one JSON object

Co-authored-by: Anthony Towns <aj@erisian.com.au>
Co-authored-by: John Newbery <john@johnnewbery.com>

MarcoFalke commented:
Addressed review from @practicalswift and @ajtowns (excellent stuff, thanks!)

In particular:
* Moved it to a new rpc to not clutter exiting rpcs
* Made the resulting dict atomic (by acquiring `cs_main`, to avoid races with incoming blocks)
* Made both types consistent (by reintroducing the "off-by-one")
* Renamed "bip9" to "vb", since BIP 8 is also a version bits deployment mechanism

I left out (from @ajtowns branch):
* 70f425e84c8d6724099a1e5f63ea309875f34848 because I didn't like the non-rpc changes. Feel free to submit as a follow up pull request
* 83709e74672d542ab58b0c70a76a32301236e0d4 because of the additional changes. Feel free to submit as follow up

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1318,29 +1309,25 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "        \"timeout\": xx,          (numeric) the median time past of a block at which the deployment is considered failed if not yet locked in\n"
             "        \"since\": xx,            (numeric) height of the first block to which the status applies\n"
-            "        \"statistics\": {         (object) numeric statistics about BIP9 signalling for a softfork (only for \"started\" status)\n"
+            "        \"statistics\": {         (object) numeric statistics about BIP9 signalling for a softfork\n"
```
Restored

Commit:
rpc: Renambe bip9 to vb in getforkinfo

ajtowns commented:
Not sure why I didn't think of it earlier, but there's a much easier way of fixing the off-by-one issue with "active" that doesn't need changes to versionbits, given we do the "height" calculation anyway:

```diff
@@ -1189,6 +1189,7 @@ static void BuriedForkDescPushBack(UniValue& softforks, const std::string& name,
     rv.pushKV("type", "buried");
     rv.pushKV("height", height);
     rv.pushKV("active", tip.nHeight + 1 >= height);
+    rv.pushKV("enforced", tip.nHeight >= height);
 
     softforks.pushKV(name, rv);
 }
@@ -1227,12 +1228,15 @@ static UniValue BIP9SoftForkDesc(const CBlockIndex* pindex, const Consensus::Par
     UniValue rv(UniValue::VOBJ);
     rv.pushKV("type", "vb");
     rv.pushKV("vb", bip9);
+    int height = -1;
     if (ThresholdState::LOCKED_IN == thresholdState) {
-        rv.pushKV("height", since_height + consensusParams.nMinerConfirmationWindow);
+        height = since_height + consensusParams.nMinerConfirmationWindow;
     } else if (ThresholdState::ACTIVE == thresholdState) {
-        rv.pushKV("height", since_height);
+        height = since_height;
     }
+    if (height >= 0) rv.pushKV("height", height);
     rv.pushKV("active", ThresholdState::ACTIVE == thresholdState);
+    rv.pushKV("enforced", (height >= 0 && pindex->nHeight >= height));
 
     return rv;
 }
```

(I added a differently named field to make sure `"active": True` matches up with `"vb": { "status": "active", .. }`, since bumping the info in the "vb" field really needs changes to versionbits.cpp to make much sense.

Also, you've got a typo in the commit message, "Renam**b**e bip9..".
